### ISSUE TITLE
Enable controllers and other elements to have multiple interfaces

### ISSUE THREAD

**Issue by buehler** [buehler/dotnet-operator-sdk#202]

**Comment by erin-allison** [comment]
I'm gonna see if I can get this put together.

**Comment by erin-allison** [comment]
Okay so the problem here seems to be that in `ResourceLocator.TypeCombo<T>`, each concrete class is only considered for its _first_ matching interface. For example, the following code segment will only have one tuple returned for `IResourceController<V1DemoEntity>, V1DemoEntity` despite implementing `IResourceController` multiple times.

```csharp
public class DemoController : IResourceController<V1DemoEntity>, IResourceController<V1DemoEntityClone>
```

**Comment by erin-allison** [comment]
My thoughts: that `ResourceLocator` class needs to be refactored out. It would resolve this...

https://github.com/buehler/dotnet-operator-sdk/blob/56e1bd3fd056e1ea4434ec43ed07df203a93b76c/src/KubeOps/Operator/Builder/OperatorBuilder.cs#L81-L82

... and also allow bringing registration behavior more in line with common standards, where we could leave the default behavior of auto-scanning the main assembly, but also allow developers to manually register controllers, finalizers, etc directly into the DI container and have the SDK pick them up and run with it.

**Comment by buehler** [comment]
Back in time I had a "registration behavior" where one could add controllers / finalizers and such elements to the kubeops - builder. Then I went to auto scanning. A mixture of both would be nice and I agree with you that the resource locator should be refactored out. I did not have time to do it though, so if you have a clever idea on how to solve the matter, I'd be happy to review it ;-)

This was version 1: https://github.com/buehler/dotnet-operator-sdk/blob/4190ec7b120a27c3f0af7b6c34f659e8b2e87d15/src/KubeOps/Operator/ServiceCollectionExtensions.cs

This was version 2: https://github.com/buehler/dotnet-operator-sdk/blob/5e0aded3cf97c6d7e44a6b87aebad232402ea0ce/src/KubeOps/Operator/Builder/IOperatorBuilder.cs

With the new structure and auto mapping I'm quite happy, but a more explicit variant could be good for the dev experience.
Furthermore, it adheres to the common standards.

**Comment by erin-allison** [comment]
I've made some decent progress so far, if you're interested in taking a peek (I'm not going to finish tonight): https://github.com/erin-allison/dotnet-operator-sdk/tree/RefactorDIRegistration

I'm going to do another decent change in how I have the registrations done, but needed to kind of re-architect things to get rid of `ResourceLocator` first.

`OperatorBuilder` now has methods like `AddController`, `AddFinalizer`, etc. and with a few small changes to the generator commands, all the tests still pass.

In the process of testing the changes on my local cluster, I think I found a few unrelated issues with the web hooks, but I will need to build my test operator against master to make sure they are unrelated.

**Comment by buehler** [comment]
Hey @erin-allison

The changes look amazing so far! I did not think about extra types back then, this is a good addition to solve the problem.

With the webhooks, what are the issues? Because they need to be built and then registered within the cluster to run. Also if you want to test webhooks locally, you'll need some software like `ngrok` or `localtunnel` to route the Kubernetes API to your local running application via HTTPS.

**Comment by erin-allison** [comment]
I don't have my notes in front of me right now but here's what I remember:
 - the RBAC role didn't include the ability to work with MutatingWebhookConfiguration
 - the RBAC role didn't include DELETE for Service or Mutating/ValidatingWebhookConfiguration (it calls delete then create in the install action
 - when I got the Mutating hook set up using the demo one in the template, it yelled at me about something else (not related to permissions) when it was called

**Comment by erin-allison** [comment]
Those are a matter for later though. That was with all my changes so there's a distinct possibility I caused at least some of my own problems.

I'll open up separate issues for those if I can reproduce against master later this weekend.

**Comment by buehler** [comment]
For the first two problems, I think I found the solution:
https://github.com/erin-allison/dotnet-operator-sdk/blob/RefactorDIRegistration/src/KubeOps/Operator/Util/RbacBuilder.cs#L49-L59

You need to add the Mutating config as well as services.Delete verb to rbac.

For the third error, I cannot give any advice.

May I ask you to put the `IRbacBuilder` into the `Rbac` directory and the `ICrdBuilder` into the `Entities` folder (and therefore namespace). I know there is a `Util` folder/namespace, but I'm not a huge fan of those. I'd rather see such implementations in a more expressing namespace than "Util".

**Comment by erin-allison** [comment]
> For the first two problems, I think I found the solution:
> https://github.com/erin-allison/dotnet-operator-sdk/blob/RefactorDIRegistration/src/KubeOps/Operator/Util/RbacBuilder.cs#L49-L59
> 
> You need to add the Mutating config as well as services.Delete verb to rbac.

Yep, that is what I was thinking, though I straight-up copy-pasted the method so I wasn't going to tweak it without further investigation.


> For the third error, I cannot give any advice.

No worries, I have the stack trace saved to debug it later.


> May I ask you to put the `IRbacBuilder` into the `Rbac` directory and the `ICrdBuilder` into the `Entities` folder (and therefore namespace). I know there is a `Util` folder/namespace, but I'm not a huge fan of those. I'd rather see such implementations in a more expressing namespace than "Util".

Yep, I agree, just wasn't 100% sure where to drop them.

**Comment by erin-allison** [comment]
Re. the issue with the webhooks, I've confirmed that blurb is exactly as it is currently on master, so fixing that will be handled separately; I don't want to try and clump in too many unrelated changes.

https://github.com/buehler/dotnet-operator-sdk/blob/bc13961d4982316fe3a1f7a2af61496677e638e8/src/KubeOps/Operator/Commands/Generators/RbacGenerator.cs#L40-L50

I've moved those items you mentioned, and I expect there may be other changes to make, but I'll mostly leave organization to the end and focus on functionality first.
